tests/Microsoft.Identity.Web.Test/WebAppExtensionsTests.cs

```diff
@@ -1,32 +1,493 @@
+        }
+
+        private void AddWebAppCallsProtectedWebApi_TestCommon(IServiceCollection services, ServiceProvider provider)
+        {
+            // Assert correct services added
+            Assert.Contains(services, s => s.ServiceType == typeof(IHttpContextAccessor));
+            Assert.Contains(services, s => s.ServiceType == typeof(ITokenAcquisition));
+            Assert.Contains(services, s => s.ServiceType == typeof(IConfigureOptions<ConfidentialClientApplicationOptions>));
+            Assert.Contains(services, s => s.ServiceType == typeof(IConfigureOptions<MicrosoftIdentityOptions>));
+            Assert.Contains(services, s => s.ServiceType == typeof(IConfigureOptions<OpenIdConnectOptions>));
+
+            // Assert OIDC options added correctly
+            var configuredOidcOptions = provider.GetService<IConfigureOptions<OpenIdConnectOptions>>() as ConfigureNamedOptions<OpenIdConnectOptions>;
+
+            Assert.Equal(_oidcScheme, configuredOidcOptions.Name);
+        }
+
+        private async Task AddWebAppCallsProtectedWebApi_TestAuthorizationCodeReceivedEvent(IServiceProvider provider, OpenIdConnectOptions oidcOptions, Func<AuthorizationCodeReceivedContext, Task> authCodeReceivedFuncMock, ITokenAcquisition tokenAcquisitionMock)
+        {
+            var httpContext = HttpContextUtilities.CreateHttpContext();
+            httpContext.RequestServices = provider;
+            var authScheme = new AuthenticationScheme(OpenIdConnectDefaults.AuthenticationScheme, OpenIdConnectDefaults.AuthenticationScheme, typeof(OpenIdConnectHandler));
+            var authProperties = new AuthenticationProperties();
+            var authCodeReceivedContext = new AuthorizationCodeReceivedContext(httpContext, authScheme, oidcOptions, authProperties);
+
+            await oidcOptions.Events.AuthorizationCodeReceived(authCodeReceivedContext).ConfigureAwait(false);
+
+            // Assert event called
+            await authCodeReceivedFuncMock.ReceivedWithAnyArgs().Invoke(Arg.Any<AuthorizationCodeReceivedContext>()).ConfigureAwait(false);
+            await ((ITokenAcquisitionInternal)tokenAcquisitionMock).ReceivedWithAnyArgs().AddAccountToCacheFromAuthorizationCodeAsync(Arg.Any<AuthorizationCodeReceivedContext>(), Arg.Any<IEnumerable<string>>()).ConfigureAwait(false);
+        }
+
+        private async Task AddWebAppCallsProtectedWebApi_TestTokenValidatedEvent(OpenIdConnectOptions oidcOptions, Func<TokenValidatedContext, Task> tokenValidatedFuncMock)
+        {
+            var httpContext = HttpContextUtilities.CreateHttpContext();
+            httpContext.Request.Form = new FormCollection(
+                new Dictionary<string, StringValues>() { { ClaimConstants.ClientInfo, Base64UrlHelpers.Encode($"{{\"uid\":\"{TestConstants.TenantIdAsGuid}\",\"utid\":\"{TestConstants.TenantIdAsGuid}\"}}") } });
```

Comment:
nit: make a specific uid constant value and utid constant value...i don't think they are normally the tenantId. but @jmprieur would probably know. And then let's assert those specific values are there below.